[dynamo] add torch._dynamo.enable and fix compile/enable/disable interaction

williamwen42 opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #132926

Fixes the compile/disable interaction part of https://github.com/pytorch/pytorch/issues/123771. Context manager to be followed-up on.

cc @ezyang @gchanan @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @chenyang78 @kadeng @chauhang @amjames

williamwen42 committed:
[dynamo] add torch._dynamo.enable

[ghstack-poisoned]

jansel commented on torch/_dynamo/decorators.py:
```diff
@@ -57,6 +57,13 @@ def disable(fn=None, recursive=True):
         return skip(fn)
 
 
+def enable(fn=None):
```
Add docstring explaining this

jansel commented on torch/_dynamo/eval_frame.py:
```diff
@@ -589,6 +589,32 @@ def __reduce__(self):
+            finally:
+                set_eval_frame_callback_enabled(prior)
+
+        _fn._torchdynamo_disable = True  # type: ignore[attr-defined]
```
Why does enable set the disable flag?  Add comment.

jansel reviewed (changes requested):
Failing tests

williamwen42 committed:
Update on "[dynamo] add torch._dynamo.enable"


Follow up to https://github.com/pytorch/pytorch/issues/123771.

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx chenyang78 kadeng chauhang amjames

[ghstack-poisoned]

anijain2305 commented on torch/_dynamo/decorators.py:
```diff
@@ -39,24 +39,71 @@ def run(fn=None):
+
+    NOTE: Interaction between `compile`, `disable`, and `enable`
+
+    `compile` is is a "marker" that Dynamo should attempt to compile the function
+    and its nested calls.
+
+    `disable` is higher-priority - it signifies that a function (and its nested
+    calls in the case recursive=True) should not be compiled.
+
+    In particular, `disable` overrides `compile` - if you want to re-enable compilation,
+    use `enable`. `disable` and `enable` have the same priority.
+
+    e.g.
+    @enable
+    def a(x):
+        ...
+
+    @disable
+    def b(x):
+        a(x)
+        ...
+
+    @compile
+    def c(x):
+        b(x)
+        ...
+
+    @disable
+    def d(x):
+        c(x)
+        ...
+
+    Calling `a`  will result in no compilation.
+    Calling `b` will result in no compilation
+    Calling `c` will result in `c` and `a` being compiled.
+    Calling `d` will result in `a` being compiled.
```
Thanks, this is very descriptive!

ezyang replied:
You probably need to say more about what graph breaks happened

williamwen42 committed:
Update on "[dynamo] add torch._dynamo.enable and fix compile/enable/disable interaction"


Follow up to https://github.com/pytorch/pytorch/issues/123771.

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx chenyang78 kadeng chauhang amjames

[ghstack-poisoned]

williamwen42 committed:
Update on "[dynamo] add torch._dynamo.enable and fix compile/enable/disable interaction"


Follow up to https://github.com/pytorch/pytorch/issues/123771.

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx chenyang78 kadeng chauhang amjames

[ghstack-poisoned]

williamwen42 committed:
Update on "[dynamo] add torch._dynamo.enable and fix compile/enable/disable interaction"


Fixes the compile/disable interaction part of https://github.com/pytorch/pytorch/issues/123771. Context manager to be followed-up on.

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx chenyang78 kadeng chauhang amjames

[ghstack-poisoned]

ezyang commented:
Is this backwards compatible?

williamwen42 commented:
No, this will be BC-breaking - see discussion in https://github.com/pytorch/pytorch/issues/123771.

ezyang commented:
I reread the thread but I couldn't tell what exactly was broken. Could you write the hypothetical release notes entry that explains the BC breaking change and how to migrate your code from old to new? (I'm still not convinced I want to do a BC break here)

williamwen42 committed:
Update on "[dynamo] add torch._dynamo.enable and fix compile/enable/disable interaction"


Fixes the compile/disable interaction part of https://github.com/pytorch/pytorch/issues/123771. Context manager to be followed-up on.

cc ezyang gchanan voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx chenyang78 kadeng chauhang amjames

[ghstack-poisoned]

williamwen42 commented:
Hypothetical release notes entry:

Added `torch.compiler.enable` and changed how `torch.compile/torch.compiler.disable/torch.compiler.enable` interact.

Before, `compile` and `disable` had the same priority and would override each other, e.g.

```python
@disable
def a(x):
    return x + 1

@compile
def b(x):
    return a(x) + 1

@disable
def c(x):
    return b(x) + 1
```

Running `a` results in no compilation. Running `b` results in `b`'s compilation. Running `c` also results in `b`'s compilation.

With `torch.compiler.enable`, `disable` now has higher priority than `compile` and can only be inverted using `enable`:
```python
    @enable
    def a(x):
        return x + 1

    @disable
    def b(x):
        return a(x) + 1

    @compile
    def c(x):
        return b(x) + 1

    @disable
    def d(x):
        return c(x) + 1
```

- Calling `a`  will result in no compilation.
- Calling `b` will result in no compilation
- Calling `c` will result in `c` and `a` being compiled. A graph break will occur when `b` is called.
- Calling `d` will result in `a` being compiled.

In order to migrate, do not use `torch.compile` as a way to invert a `torch.compiler.disable`. Use `torch.compiler.enable` instead. It is possible to wrap `compile` with `enable` if the `compile` call is not known to be the top level `compile` call or not.

For example:
```python
@compile 
def a(x):
    return x + 1

@disable
def b(x):
    return a(x) + 1

@compile
def c(x):
    return b(x) + 1
```
`a` can be migrated to one of:
```python
@enable
def a(x):
    return x + 1

@enable
@compile
def a(x):
    return x + 1
```

williamwen42 committed:
Update on "[dynamo] add torch._dynamo.enable and fix compile/enable/disable interaction"


Fixes the compile/disable interaction part of https://github.com/pytorch/pytorch/issues/123771. Context manager to be followed-up on.

cc ezyang gchanan voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx chenyang78 kadeng chauhang amjames

[ghstack-poisoned]

williamwen42 committed:
Update on "[dynamo] add torch._dynamo.enable and fix compile/enable/disable interaction"


Fixes the compile/disable interaction part of https://github.com/pytorch/pytorch/issues/123771. Context manager to be followed-up on.

cc ezyang gchanan voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx chenyang78 kadeng chauhang amjames

[ghstack-poisoned]

ezyang reviewed (approved):
Yes, note is very clear, thanks!

ezyang commented:
Because this is BC breaking you need a kill switch for this PR, in case it breaks someone internally

williamwen42 committed:
Update on "[dynamo] add torch._dynamo.enable and fix compile/enable/disable interaction"


Fixes the compile/disable interaction part of https://github.com/pytorch/pytorch/issues/123771. Context manager to be followed-up on.

cc ezyang gchanan voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx chenyang78 kadeng chauhang amjames

[ghstack-poisoned]

williamwen42 committed:
Update on "[dynamo] add torch._dynamo.enable and fix compile/enable/disable interaction"


Fixes the compile/disable interaction part of https://github.com/pytorch/pytorch/issues/123771. Context manager to be followed-up on.

cc ezyang gchanan voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx chenyang78 kadeng chauhang amjames

[ghstack-poisoned]

williamwen42 commented:
@ezyang added a killswitch - check it over?

williamwen42 committed:
Update on "[dynamo] add torch._dynamo.enable and fix compile/enable/disable interaction"


Fixes the compile/disable interaction part of https://github.com/pytorch/pytorch/issues/123771. Context manager to be followed-up on.

cc ezyang gchanan voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx chenyang78 kadeng chauhang amjames

[ghstack-poisoned]

ezyang commented on torch/_dynamo/config.py:
```diff
@@ -374,6 +374,12 @@ def _get_optimize_ddp_mode():
+# where compile and disable have the same priority.
+# Killswitch in case internal breaks.
+# See https://github.com/pytorch/pytorch/pull/132926.
+old_compile_disable_behavior = False
```
This isn't a useful killswitch. You haven't hooked up to a JK (which we can use to change this value without a code change), and your value needs to default to True (because that's how JKs work, they default to True)

williamwen42 committed:
Update on "[dynamo] add torch._dynamo.enable and fix compile/enable/disable interaction"


Fixes the compile/disable interaction part of https://github.com/pytorch/pytorch/issues/123771. Context manager to be followed-up on.

cc ezyang gchanan voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx chenyang78 kadeng chauhang amjames

[ghstack-poisoned]

williamwen42 committed:
Update on "[dynamo] add torch._dynamo.enable and fix compile/enable/disable interaction"


Fixes the compile/disable interaction part of https://github.com/pytorch/pytorch/issues/123771. Context manager to be followed-up on.

cc ezyang gchanan voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx chenyang78 kadeng chauhang amjames

[ghstack-poisoned]

williamwen42 commented:
@ezyang added JK killswitch

ezyang commented on torch/utils/_content_store.py:
```diff
@@ -53,7 +53,7 @@ def lazy_compile(**compile_kwargs):
         @functools.wraps(fn)
         def compile_hook(*args, **kwargs):
-            compiled_fn = torch.compile(fn, **compile_kwargs)
+            compiled_fn = torch.compiler.enable(torch.compile(fn, **compile_kwargs))
```
This is interesting, where the default behavior you want for disable interaction is different for "I'm just using torch.compile to enable compilation for a single operator with dynamic shapes". Maybe we need to deify this into a real API / give it a different name so it's more obviously different from compile?

williamwen42 replied:
Resolved to not add a new API for this.

ezyang commented:
Looking over some of the breakages in CI, I now think we need to actually come up with a detailed BC plan.

The crux of the problem is sometimes, people use torch.compile to mean "I just want to compile this little fragment of code as an internal implementation detail, so I don't have to write an eager CUDA kernel for it." These uses of torch.compile should be indifferent to an enclosing disable, because they're not compile in the conventional sense; they are an implementation detail that shouldn't be affected by this API.

With the change here, you're not going to notice that you've broken people, unless your end use cases happen to be mandatory torch.compile (as was the case with content store and flex attention in the API). Other people relying on this will just see their little fragments just silently stop compiling, someone will need to go in and diff before/after to find out the cause of the QPS difference.

Maybe there is a way we can still land something like this, but I'm not seeing an obvious way to do it. @Chillee any more thoughts...

williamwen42 commented:
@Chillee @ezyang follow up?

ezyang commented:
@williamwen42 did you have any ideas?

williamwen42 commented:
Was wondering if @Chillee had any thoughts since he wanted the new compile/disable interaction but there are cases where the old behavior is expected.

ezyang commented:
I'm kind of liking an API for mandatory compile cases. It would help that other PR you opened too, where you noted that it could impede "operator-like" use of torch.compile. It could be something as simple as slapping enable on operator-like usages, no new API needed.

ezyang commented:
d'oh, which is what you did in this PR. Maybe we just need a higher level API then? There seem to be two primary use cases right now: (1) operator like usage of compile (content store) and (2) higher order ops. Paved APIs for "hey, always compile me" in both cases would make it more likely users would use them. Do we also need to do a study of the BC impact of this internally in fbcode? Not sure, I'll leave that to your discretion.

williamwen42 commented:
Offline discussion with @anijain2305: Consider adding a warning clarifying compile/disable/enable semantics for a few releases when we detect `@compile` under a `@disable`.

Offline discussion with @Chillee: Doubtful that many users (esp. internal) depend on current compile/disable semantics - no strong reason why one should rely on these undocumented semantics. We can just go ahead and if things really break internally, we can revert/killswitch. Having a new API (e.g. `force_compile/always_compile`) being an alias for `@enable @compile` may not be a great idea since it would involve adding new user-facing API that needs to be maintained.

williamwen42 committed:
Update on "[dynamo] add torch._dynamo.enable and fix compile/enable/disable interaction"


Fixes the compile/disable interaction part of https://github.com/pytorch/pytorch/issues/123771. Context manager to be followed-up on.

cc ezyang gchanan voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx chenyang78 kadeng chauhang amjames

[ghstack-poisoned]

williamwen42 committed:
Update on "[dynamo] add torch._dynamo.enable and fix compile/enable/disable interaction"


Fixes the compile/disable interaction part of https://github.com/pytorch/pytorch/issues/123771. Context manager to be followed-up on.

cc ezyang gchanan voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx chenyang78 kadeng chauhang amjames

[ghstack-poisoned]

williamwen42 commented:
Updates:
- Added `force=True` option to `torch.compile` in order to force compilation in a disable region
- Issue a warning if compile is attempted without `force=True` in a disable region.

ezyang commented:
Do you like force=True more than just putting an enable decorator on top? They both seem bad so I'm kind of leaning towards not adding more API surface

williamwen42 commented:
Right now, the issue with just using an enable decorator is that we issue a warning (when from a UX perspective we shouldn't) if the order of `compile` and `enable` is incorrect (e.g. `b = torch.compile(torch.compiler.enable(lambda x: x + 1), backend="eager")` will issue a warning).

Using a new API will fix this ordering issue, although we could forego the new API and check in `compile` to see if we're wrapping an `enable` call.

ezyang commented:
Concretely, I suggest *never* raising a warning here!

williamwen42 committed:
Update on "[dynamo] add torch._dynamo.enable and fix compile/enable/disable interaction"


Fixes the compile/disable interaction part of https://github.com/pytorch/pytorch/issues/123771. Context manager to be followed-up on.

cc ezyang gchanan voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx chenyang78 kadeng chauhang amjames

[ghstack-poisoned]

williamwen42 commented:
@ezyang removed the warning and the API. Are we ok landing this (given tests pass)?

ezyang commented on torch/_dynamo/config.py:
```diff
@@ -382,6 +382,13 @@ def _get_optimize_ddp_mode():
+# where compile and disable have the same priority.
+# Killswitch in case internal breaks.
+# See https://github.com/pytorch/pytorch/pull/132926.
+old_compile_disable_behavior = False
```
@c00w is your thing ready to be used here?

ezyang commented on torch/_dynamo/decorators.py:
```diff
@@ -47,9 +53,17 @@ def run(fn=None):
 
+def _old_compile_disable_behavior():
+    return config.old_compile_disable_behavior or not justknobs_check(
+        "pytorch/compiler:compile_enable_disable"
```
you don't have to fix it now, but this is a pretty bad JK name; why doesn't it have "old" or "behavior" in its name? Is it enabling something or disabling something?

ezyang replied:
How are you testing the alternate JK setting :think:

williamwen42 replied:
Added a `test_old_compile_disable` test that sets `config.old_compile_disable_behavior` to true.

ezyang commented on torch/compiler/__init__.py:
```diff
@@ -216,18 +217,69 @@ def assume_constant_result(fn):
+    """
+    Decorator to re-enable compilation - inverse of `disable`.
+
+    Compilation will only occur if there was a previous `compile` call.
```
I think this docstring can be better. At minimum, it can explicitly discuss some of the other use cases we have for enable; e.g., if you have a kernel that you want always compiled because you're using PT2 for kernel authoring

williamwen42 replied:
Added always-compile kernel example to docs

ezyang commented on torch/utils/_content_store.py:
```diff
@@ -53,7 +53,7 @@ def lazy_compile(**compile_kwargs):
     def decorate_fn(fn):
```
It's probably worth updating the doc here too, since this isn't strictly a lazy compile only anymore; it also forces compilation (so it's really a wrapper oriented towards authoring eager kernels)

ezyang commented on torch/nn/attention/flex_attention.py:
```diff
@@ -846,7 +846,9 @@ def causal_mask(b, h, q_idx, kv_idx):
     if _compile:
-        inner_func = torch.compile(inner_func, fullgraph=True, dynamic=False)
+        inner_func = torch.compile(
+            torch.compiler.enable(inner_func), fullgraph=True, dynamic=False
```
Err, don't you need to enable before compile?

ezyang replied:
Oh, actually, I think it doesn't matter? compile and enable are commutative, aren't they...

williamwen42 replied:
Only matters if we end up issuing a warning.

ezyang commented on torch/compiler/__init__.py:
```diff
@@ -216,18 +217,69 @@ def assume_constant_result(fn):
+        Calling `b` will result in no compilation
+        Calling `c` will result in `c` and `a` being compiled.
+            A graph break will occur when `b` is called.
+        Calling `d` will result in `a` being compiled.
```
Maybe we should also talk a little bit about disable recursive=False.  Non-recursive disable is equivalent to immediately "re-enabling" PT2 on all the subcalls right? (Is this what actually happens?)

williamwen42 replied:
skip/recursive=False is implemented by applying `SKIP_CODE` to the current frame and marking the function with `_torchdynamo_disable`, which causes graph breaks if we attempt to inline. It does not touch `eval_frame_callback_enabled`.

There are some behaviors of skip/recursive=False that are not immediately obvious:
- subcalls are not actively re-enabled. If we are in a recursive=True disable region, subcalls continue to be disabled.
- nested compile/enable/disable decorators are skipped.

I think the simplest solution is to declare recursive=False as even higher priority than recursive=True/enable. Because recursive=False only applies to one frame, there is no need to "re-enable".

ezyang commented on torch/csrc/dynamo/eval_frame.c:
```diff
@@ -18,6 +18,7 @@ static char compile_context[MAX_COMPILE_CONTEXT_SIZE];
 static int active_dynamo_threads = 0;
 
 static Py_tss_t eval_frame_callback_key = Py_tss_NEEDS_INIT;
+static bool eval_frame_callback_enabled = true;
```
Are you sure this doesn't need to be thread local 🤔

williamwen42 replied:
I think `eval_frame_callback_key ` and `eval_frame_callback_enabled ` storage class specifiers should match since these two vars are used in similar ways. Would not using `thread_local` here lead to bugs right now, or are we future-proofing for multithreaded dynamo?

ezyang commented on torch/csrc/dynamo/eval_frame.c:
```diff
@@ -509,7 +510,7 @@ static PyObject* _custom_eval_frame_shim(
 
-  if (callback == Py_None) {
+  if (!eval_frame_callback_enabled || callback == Py_None) {
     return eval_frame_default(tstate, frame, throw_flag);
```
Would it make more sense for eval_frame_callback_get to consult the enable/disable status?

williamwen42 replied:
I want to keep `eval_frame_callback_enabled` and `eval_frame_callback_get` independent. `eval_frame_callback_get` should return the callback dynamo _wants_ to use, `eval_frame_callback_enabled` determines if we actually use the callback or the default.

In particular, `set_eval_frame_py` (used in Dynamo convert_frame) retrieves the previous set callback via `eval_frame_callback_get()`, and I don't want `eval_frame_callback_enabled` getting in the way.

ezyang commented:
> Right now, the issue with just using an enable decorator is that we issue a warning (when from a UX perspective we shouldn't) if the order of compile and enable is incorrect (e.g. b = torch.compile(torch.compiler.enable(lambda x: x + 1), backend="eager") will issue a warning).

So, I was reflecting on this more, and I think there's a variant of this warning we should raise.

Specifically, if I have:

```
@disable
def f():
  g()

@torch.compile()
def g():
  ...
```

When we call f, I think it would be very reasonable for g to raise a warning (once!) saying that you requested torch.compile, but it was disabled by higher frame, so it's not enabled. And suggest wrapping in enable to force compilation. @Chillee do you think such a warning is reasonable, or annoying? I am willing to be talked out of it.

Assuming we do have a warning, though, then, the preferred ordering for enable + compile is wrap enable AROUND compile, since that will prevent a warning.

ezyang commented on torch/_dynamo/decorators.py:
```diff
@@ -58,15 +72,35 @@ def disable(fn=None, recursive=True):
         if fn is not None:
-            fn = innermost_fn(fn)
+            if _old_compile_disable_behavior():
+                fn = innermost_fn(fn)
```
What's going on here?

williamwen42 replied:
The previous behavior would skip nested compile/disable decorators.

ezyang commented on torch/compiler/__init__.py:
```diff
@@ -216,18 +217,69 @@ def assume_constant_result(fn):
+        and its nested calls.
+
+        `disable` is higher-priority - it signifies that a function (and its nested
+        calls in the case recursive=True) should not be compiled.
```
I think that we should explicitly describe the model where there is a flag (is compile enabled/disabled) and then describe these API functions in terms of the flag.

ezyang commented on torch/_dynamo/decorators.py:
```diff
@@ -58,15 +72,35 @@ def disable(fn=None, recursive=True):
+    """
+    Decorator to re-enable TorchDynamo - inverse of `disable`.
+
+    Compilation will only occur if there was a previous `compile` call.
```
Are your canonical docs going to live here or in torch.compiler?

williamwen42 replied:
torch.compiler

ezyang commented on torch/_dynamo/eval_frame.py:
```diff
@@ -645,6 +648,55 @@ def __reduce__(self):
+            # if dynamo is already enabled, use _torchdynamo_disable to skip going through _fn
+            _fn._torchdynamo_disable = True  # type: ignore[attr-defined]
+
+        return _fn
```
This seems way more complicated than I expected it to be. In particular, why do you need to apply OptimizedModule?

ezyang replied:
I get a lot of "copy paste" energy here (maybe a lot of this code was preexisting?) But then, why is it being copy pasted?

williamwen42 replied:
> why do you need to apply OptimizedModule
So that the result of enable/disable can be a module if the original function is a module. e.g.
```python
mod1.layer = torch.compiler.enable(mod2)
```

> why is it being copy pasted?
Copy pasted from the original `DisableContext` so that I can preserve a copy to alleviate BC-breaking concerns. I could merge these 2 (`OldDisableContext` and `CompileEnabledContext`) together.

ezyang replied:
I think I definitely prefer having all of the OptimizedModule wrapping logic together, although I do think there's some implementer's discretion here.

williamwen42 replied:
Eventually I'd like to get rid of `OldDisableContext`, so keeping a separate copy makes it easy to remove.

williamwen42 commented:
@ezyang regarding issuing the warning, I believe the previous version of this PR with the warning implements what you are thinking of (only issue a warning once per frame if we detect a compile call within a disable region).

ezyang commented:
Ah, this is a misunderstanding then. Specifically:

> Right now, the issue with just using an enable decorator is that we issue a warning (when from a UX perspective we shouldn't) if the order of compile and enable is incorrect (e.g. b = torch.compile(torch.compiler.enable(lambda x: x + 1), backend="eager") will issue a warning).

You proposed fixing this by adding another flag to suppress the warning (which I didn't like), but I didn't realize that you can just interchange the compile-enable to suppress the warning (no new flag needed).

ezyang reviewed (changes requested):
I'm pretty sure the TSS handling is not right

williamwen42 committed:
Update on "[dynamo] add torch._dynamo.enable and fix compile/enable/disable interaction"


Fixes the compile/disable interaction part of https://github.com/pytorch/pytorch/issues/123771. Context manager to be followed-up on.

cc ezyang gchanan voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx chenyang78 kadeng chauhang amjames

[ghstack-poisoned]

williamwen42 committed:
Update on "[dynamo] add torch._dynamo.enable and fix compile/enable/disable interaction"


Fixes the compile/disable interaction part of https://github.com/pytorch/pytorch/issues/123771. Context manager to be followed-up on.

cc ezyang gchanan voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx chenyang78 kadeng chauhang amjames

[ghstack-poisoned]

williamwen42 committed:
Update on "[dynamo] add torch._dynamo.enable and fix compile/enable/disable interaction"


Fixes the compile/disable interaction part of https://github.com/pytorch/pytorch/issues/123771. Context manager to be followed-up on.

cc ezyang gchanan voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx chenyang78 kadeng chauhang amjames

[ghstack-poisoned]

williamwen42 committed:
Update on "[dynamo] add torch._dynamo.enable and fix compile/enable/disable interaction"


Fixes the compile/disable interaction part of https://github.com/pytorch/pytorch/issues/123771. Context manager to be followed-up on.

cc ezyang gchanan voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx chenyang78 kadeng chauhang amjames

[ghstack-poisoned]

williamwen42 committed:
Update on "[dynamo] add torch._dynamo.enable and fix compile/enable/disable interaction"


Fixes the compile/disable interaction part of https://github.com/pytorch/pytorch/issues/123771. Context manager to be followed-up on.

cc ezyang gchanan voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx chenyang78 kadeng chauhang amjames

[ghstack-poisoned]

williamwen42 committed:
Update on "[dynamo] add torch._dynamo.enable and fix compile/enable/disable interaction"


Fixes the compile/disable interaction part of https://github.com/pytorch/pytorch/issues/123771. Context manager to be followed-up on.

cc ezyang gchanan voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx chenyang78 kadeng chauhang amjames

[ghstack-poisoned]

williamwen42 committed:
Update on "[dynamo] add torch._dynamo.enable and fix compile/enable/disable interaction"


Fixes the compile/disable interaction part of https://github.com/pytorch/pytorch/issues/123771. Context manager to be followed-up on.

cc ezyang gchanan voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx chenyang78 kadeng chauhang amjames

[ghstack-poisoned]

ezyang commented on torch/compiler/__init__.py:
```diff
@@ -216,18 +217,110 @@ def assume_constant_result(fn):
+    .. note::
+        Interaction between ``compile``, recursive ``disable``, and ``enable``
+
+        ``compile`` is is a "marker" that Dynamo should attempt to compile the function
```
s/is is/is/

ezyang commented on torch/compiler/__init__.py:
```diff
@@ -216,18 +217,110 @@ def assume_constant_result(fn):
+
+    Compilation will only occur if there was a previous `compile` call.
+
+    `enable` is useful for finegrained compiilation control. See the note in `disable` for
```
s/compiilation/compilation/

ezyang commented:
Hi, sorry, I'm going to thrash you again.

I was reviewing the docs you added and proposing a rewrite of the docs, and while doing so, I realized that I actually think disable as it is used today is different from the "disable" that Horace was proposing in the issue, and we should actually give it a different name. Horace and I came up with "torch.compiler.bypass", but Claude also suggested "force_uncompiled" and "prevent_compilation" which also seemed good. Here are my proposed docs for disable vs these other options.

----

What is the difference between torch.compiler.disable and torch.compiler.force_uncompiled? It's easiest to consider the use cases for these two functions.

torch.compiler.disable is for when you want to prevent an outer compilation from continuing compiling into an inner function. You typically want to do this if you have some code that is reachable from a torch.compile that triggers a problem in the compiler (e.g., it causes infinite recompilations) and you just want to disable compilation for it. It comes in two flavors: the default is to disable compilation for that function and all recursive calls (unless you explicitly call torch.compile again internally), and recursive=False, where it will only disable compilation for that particular function call, but will attempt to continue compilation for inner function calls.

torch.compiler.force_uncompiled is for when you are in an outer function, and you want to prevent *inner* compilations from occurring. It's useful if you have some code that has torch.compile directly annotated on functions, but you want to temporarily disable the compiler / any preexisting compiled code so that you can run them under eager semantics, e.g., for benchmarking or if you want to run a TorchDispatchMode (like FlopCounter) which only works in eager semantics. If necessary, you can also undo the effect of this context manager by saying force_uncompiled(False). force_uncompiled is forbidden in graph; you're not allowed to compile code that has a force_uncompiled invocation in it.

----

I don't think you need to change much in your PR. Here's what changes:

1. The PR is no longer BC breaking. We don't need a JK, we don't need to annotate every HOP with enable. There is a funny interaction with HOPs and FlopCounterMode but we can deal with it when we get there. `force_uncompiled` won't work too well with things that require mandatory compilation, so we could still have a `force_uncompiled(False)` decorator on the content store helper.
2. force_uncompiled can be a context manager, because it doesn't rely on setting up an eval frame handler to work, unlike torch.compile. Consequently, it's not necessary to have OptimizedModule logic for it; just do the context manager syntax instead if you need to work with a module.
3. You need to rename your disable implementation, but all the other pieces stay the same. In particular, this controlling TLS that is accessed by eval frame handler still seems like the right implementation strategy to me.

WDYT?

ezyang commented:
We talked about this today in more general at https://docs.google.com/document/d/1qlQyjbENz45L7uGglmiF6CkRgr_VOdxFAcZGCqdtLTY/edit

ezyang commented:
> SGTM from an implementation perspective, although I'm concerned about sufficiently differentiating between disable and force_uncompiled. It's still not clear to me from a user perspective as to when I should use one over the other.

It was pretty clear to me when I was writing the docs, so I guess the new docs are not good enough 🤣 

Here's a shorter pitch: if you are currently *in* a torch.compile, but you want to get out, use `torch.compiler.disable`. If you are *not* in a torch.compile, but you don't want to compile for any of your nested function calls, use `torch.compiler.set_phase("force_eager")` (my new name for the API LOL).

> are you suggesting that force_uncompiled should only be a context manager? The function/decorator version is very close in effect to disable, but has small but important differences.

I don't mind a decorator but it should be a contextlib.contextmanager style decorator. I don't think `set_phase` needs to support OptimizedModule.

> "force_uncompiled is forbidden in graph" - do you mean we should skip the frame if we encounter the context manager while tracing? Or graph break if we encounter?

"forbidden in graph" is a preexisting concept which makes the compiler hard error if it traverses it while tracing. It is saying "this code should never be compiled and if you tried to compile it some big accident happened so raise a major error."

williamwen42 commented:
Difference between disable and force_eager makes more sense now.

What other options are you thinking for `set_phase`? As of now, it looks like there should be at least a `force_eager` and a default option?

ezyang commented:
The ones we have right now are: "force_eager", "default", "fail_on_recompile", "eager_on_recompile"

ezyang commented:
And I will happily bikeshed this forever if needed

yf225 commented:
To add more color from previous discussion with Ed and others, summary of motivations: https://docs.google.com/document/d/1qlQyjbENz45L7uGglmiF6CkRgr_VOdxFAcZGCqdtLTY/edit#bookmark=id.nl59a72dv4o. Particularly `torch.compiler.set_phase("force_eager")` will be super useful for Traceable FSDP / Compiled DDP / float8 / optimizer warmup phase, and the recompile-related options will also be helpful.

williamwen42 commented:
Moved to a new PR: https://github.com/pytorch/pytorch/pull/137504
